composer/callbacks/image_monitor.py

```diff
@@ -0,0 +1,194 @@
+        self.target_key = target_key
+
+        try:
+            import wandb
+        except ImportError as e:
+            raise MissingConditionalImportError(extra_deps_group='wandb',
+                                                conda_package='wandb',
+                                                conda_channel='conda-forge') from e
+        del wandb  # unused
+
+        # Check that the output mode is valid
+        if self.mode.lower() not in ['input', 'segmentation']:
+            raise ValueError(f'Invalid mode: {mode}')
+
+        # Check that the interval timestring is parsable and convert into time object
+        try:
+            self.interval = Time.from_timestring(interval)
+        except ValueError as error:
+            raise ValueError(f'Invalid time string for parameter interval') from error
+
+        # Verify that the interval has supported units.
+        if self.interval.unit not in [TimeUnit.BATCH, TimeUnit.EPOCH]:
+            raise ValueError(f'Invalid time unit for parameter interval: '
+                             f'{self.interval.unit}')
+
+    def _log_inputs(self, state: State, logger: Logger, key: str):
+        inputs = state.batch_get_item(key=self.input_key)
+        if not isinstance(inputs, torch.Tensor):
+            raise NotImplementedError('Multiple input tensors not supported yet')
+        # Verify inputs is a valid shape for conversion to an image
+        if _check_for_image_format(inputs):
+            table = _make_input_images(inputs, self.num_images)
+            logger.data_batch({key: table})
+
+    def _log_segmented_inputs(self, state: State, logger: Logger, key: str):
+        inputs = state.batch_get_item(key=self.input_key)
+        targets = state.batch_get_item(key=self.target_key)
+        outputs = state.outputs
+
+        if not isinstance(inputs, torch.Tensor):
+            raise NotImplementedError('Multiple input tensors not supported yet')
+        if not isinstance(targets, torch.Tensor):
+            raise NotImplementedError('Multiple target tensors not supported yet')
+        if not isinstance(outputs, torch.Tensor):
+            raise NotImplementedError('Multiple output tensors not supported yet')
+
+        table = _make_segmentation_images(inputs, targets, outputs, self.num_images, self.shift_targets)
+        logger.data_batch({key: table})
+
+    def before_forward(self, state: State, logger: Logger):
+        assert isinstance(self.interval, Time)
+        if self.mode.lower() == 'input' and state.timestamp.get(self.interval.unit).value % self.interval.value == 0:
+            self._log_inputs(state, logger, 'Images/Train')
+
+    def eval_before_forward(self, state: State, logger: Logger):
+        assert isinstance(self.interval, Time)
+        if self.mode.lower() == 'input' and state.eval_timestamp.get(TimeUnit.BATCH).value == 0:
+            self._log_inputs(state, logger, 'Images/Eval')
+
+    def before_loss(self, state: State, logger: Logger):
```

Comment:
Nit: is there a reason you chose BEFORE_LOSS instead of AFTER_FORWARD? It seems like no events happen in between them. Is there an algorithm that could run at AFTER_FORWARD that we want to run first? Maybe for clarity, we could do AFTER_FORWARD since we do EVAL_AFTER_FORWARD unless there is a reason not to.

Reply:
I chose BEFORE_LOSS because I want to log the targets as the loss function sees them, so this seems safest to me. Some algorithms (cutmix for example) modify the targets on the BEFORE_LOSS event, so logging images on AFTER_FORWARD here wouldn't pick up those changes.

Reply:
ok makes sense. thanks!